### PR TITLE
feat: align ExportImportedSpecifierDependency init fragments

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
## Summary
1. Closed https://github.com/web-infra-dev/rspack/issues/4935
2. Eliminated `es` globalRuntime in ExportImportedSpecifierDependency
3. Now the harmonyExportMap is  sorted see https://github.com/web-infra-dev/rspack/pull/4995/files#r1427600363 

## Require Documentation?

- [ ] No
- [ ] Yes, the corresponding rspack-website PR is \_\_

### CODE REVIEW

File: crates/rspack/tests/tree-shaking/explicit_named_export_higher_priority_1/snapshot/new_treeshaking.snap [hunk 1]
```diff
@@ -15,7 +15,6 @@ __webpack_require__.d(__webpack_exports__, {
   a: function() { return a; }
 });
 /* harmony import */var _bar__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./bar */"./bar.js");
-__webpack_require__.es(_bar__WEBPACK_IMPORTED_MODULE_0__, __webpack_exports__);
```
Comment (IWANABETHATGUY) [inline crates/rspack/tests/tree-shaking/explicit_named_export_higher_priority_1/snapshot/new_treeshaking.snap:18]: align with webpack: 
image

File: crates/rspack/tests/tree-shaking/conflicted_name_by_re_export_all_should_be_hidden/snapshot/new_treeshaking.snap [hunk 2]
```diff
@@ -18,9 +18,7 @@ __webpack_require__.r(__webpack_exports__);
 "use strict";
 __webpack_require__.r(__webpack_exports__);
 /* harmony import */var _foo__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./foo */"./foo.js");
-__webpack_require__.es(_foo__WEBPACK_IMPORTED_MODULE_0__, __webpack_exports__);
 /* harmony import */var _bar__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(/*! ./bar */"./bar.js");
-__webpack_require__.es(_bar__WEBPACK_IMPORTED_MODULE_1__, __webpack_exports__);
```
Comment (IWANABETHATGUY) [inline crates/rspack/tests/tree-shaking/conflicted_name_by_re_export_all_should_be_hidden/snapshot/new_treeshaking.snap:23]: align with webpack, 
image

File: crates/rspack/tests/tree-shaking/export-star-chain/snapshot/new_treeshaking.snap [hunk 3]
```diff
@@ -22,19 +22,23 @@ __webpack_require__.d(__webpack_exports__, {
 "./colors/c.js": (function (__unused_webpack_module, __webpack_exports__, __webpack_require__) {
 "use strict";
 __webpack_require__.r(__webpack_exports__);
+__webpack_require__.d(__webpack_exports__, {
```
Comment (IWANABETHATGUY) [inline crates/rspack/tests/tree-shaking/export-star-chain/snapshot/new_treeshaking.snap:25]: image

File: crates/rspack/tests/tree-shaking/export-star-chain/snapshot/new_treeshaking.snap [hunk 4]
```diff
@@ -22,19 +22,23 @@ __webpack_require__.d(__webpack_exports__, {
...
+  result: function() { return /* reexport safe */ _result__WEBPACK_IMPORTED_MODULE_0__.result; }
+});
 /* harmony import */var _result__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./result */"./colors/result.js");
-__webpack_require__.es(_result__WEBPACK_IMPORTED_MODULE_0__, __webpack_exports__);
 
 }),
 "./colors/index.js": (function (__unused_webpack_module, __webpack_exports__, __webpack_require__) {
 "use strict";
 __webpack_require__.r(__webpack_exports__);
+__webpack_require__.d(__webpack_exports__, {
```
Comment (IWANABETHATGUY) [inline crates/rspack/tests/tree-shaking/export-star-chain/snapshot/new_treeshaking.snap:34]: image

File: crates/rspack/tests/tree-shaking/export_star/snapshot/new_treeshaking.snap [hunk 5]
```diff
@@ -7,12 +7,12 @@ source: crates/rspack_testing/src/run_fixture.rs
 "use strict";
 __webpack_require__.r(__webpack_exports__);
 __webpack_require__.d(__webpack_exports__, {
-  bar: function() { return _foo__WEBPACK_IMPORTED_MODULE_0__; },
+  bar: function() { return /* reexport module object */ _foo__WEBPACK_IMPORTED_MODULE_0__; },
```
Comment (IWANABETHATGUY) [inline crates/rspack/tests/tree-shaking/export_star/snapshot/new_treeshaking.snap:10]: image

File: crates/rspack/tests/tree-shaking/export_star/snapshot/new_treeshaking.snap [hunk 6]
```diff
@@ -21,13 +21,14 @@ __webpack_require__.es(_result__WEBPACK_IMPORTED_MODULE_1__, __webpack_exports__
 "use strict";
 __webpack_require__.r(__webpack_exports__);
 __webpack_require__.d(__webpack_exports__, {
+  bar: function() { return /* reexport safe */ _bar__WEBPACK_IMPORTED_MODULE_0__.bar; },
```
Comment (IWANABETHATGUY) [inline crates/rspack/tests/tree-shaking/export_star/snapshot/new_treeshaking.snap:24]: image

File: crates/rspack/tests/tree-shaking/inherit_export_map_should_lookup_in_dfs_order/snapshot/new_treeshaking.snap [hunk 7]
```diff
@@ -21,12 +21,12 @@ __webpack_require__.r(__webpack_exports__);
 "use strict";
 __webpack_require__.r(__webpack_exports__);
 __webpack_require__.d(__webpack_exports__, {
+  b: function() { return /* reexport safe */ _foo__WEBPACK_IMPORTED_MODULE_0__.b; },
+  c: function() { return /* reexport safe */ _foo__WEBPACK_IMPORTED_MODULE_0__.c; },
```
Comment (IWANABETHATGUY) [inline crates/rspack/tests/tree-shaking/inherit_export_map_should_lookup_in_dfs_order/snapshot/new_treeshaking.snap:25]: image

File: crates/rspack/tests/tree-shaking/inherit_export_map_should_lookup_in_dfs_order/snapshot/new_treeshaking.snap [hunk 8]
```diff
@@ -35,10 +35,10 @@ __webpack_require__.es(_bar__WEBPACK_IMPORTED_MODULE_1__, __webpack_exports__);
 "use strict";
 __webpack_require__.r(__webpack_exports__);
 __webpack_require__.d(__webpack_exports__, {
+  c: function() { return /* reexport safe */ _a_js__WEBPACK_IMPORTED_MODULE_0__.c; },
```
Comment (IWANABETHATGUY) [inline crates/rspack/tests/tree-shaking/inherit_export_map_should_lookup_in_dfs_order/snapshot/new_treeshaking.snap:38]: image

File: crates/rspack/tests/tree-shaking/transitive-bailout/snapshot/new_treeshaking.snap [hunk 9]
```diff
@@ -29,8 +29,11 @@ _answer__WEBPACK_IMPORTED_MODULE_0__.test;
 "./lib.js": (function (__unused_webpack_module, __webpack_exports__, __webpack_require__) {
 "use strict";
 __webpack_require__.r(__webpack_exports__);
+__webpack_require__.d(__webpack_exports__, {
+  b: function() { return /* reexport safe */ _a_js__WEBPACK_IMPORTED_MODULE_0__.b; },
+  a: function() { return /* reexport safe */ _a_js__WEBPACK_IMPORTED_MODULE_0__.a; }
```
Comment (IWANABETHATGUY) [inline crates/rspack/tests/tree-shaking/transitive-bailout/snapshot/new_treeshaking.snap:34]: image

**Comment by IWANABETHATGUY** [comment]
!canary

File: crates/rspack_core/src/init_fragment.rs [hunk 10]
```diff
@@ -274,10 +276,10 @@ impl HarmonyExportInitFragment {
 }
 
 impl<C: InitFragmentRenderContext> InitFragment<C> for HarmonyExportInitFragment {
-  fn contents(self: Box<Self>, context: &mut C) -> Result<InitFragmentContents> {
+  fn contents(mut self: Box<Self>, context: &mut C) -> Result<InitFragmentContents> {
     context.add_runtime_requirements(RuntimeGlobals::EXPORTS);
     context.add_runtime_requirements(RuntimeGlobals::DEFINE_PROPERTY_GETTERS);
-
+    self.export_map.sort_by(|a, b| a.0.cmp(&b.0));
```
Comment (IWANABETHATGUY) [inline crates/rspack_core/src/init_fragment.rs:282]: https://github.com/webpack/webpack/blob/1f99ad6367f2b8a6ef17cce0e058f7a67fb7db18/lib/dependencies/HarmonyExportInitFragment.js#L149-L151

File: crates/rspack_plugin_javascript/src/dependency/commonjs/commonjs_export_require_dependency.rs [hunk 11]
```diff
@@ -0,0 +1,89 @@
+use rspack_core::{
```
Comment (IWANABETHATGUY) [inline crates/rspack_plugin_javascript/src/dependency/commonjs/commonjs_export_require_dependency.rs:1]: A very basic implementation of export_require_dependency, full impl with optimization need parser hook

Comment (IWANABETHATGUY) [reply crates/rspack_plugin_javascript/src/dependency/commonjs/commonjs_export_require_dependency.rs:1]: alread create a new issue to track this

File: crates/rspack_plugin_javascript/src/visitors/dependency/common_js_export_scanner.rs [hunk 12]
```diff
@@ -182,6 +185,32 @@ impl Visit for CommonJsExportDependencyScanner<'_> {
             // this = require('xx');
             // It's possible to reexport __esModule, so we must convert to a dynamic module
             self.set_dynamic();
+            let related_require_dep = self
```
Comment (IWANABETHATGUY) [inline crates/rspack_plugin_javascript/src/visitors/dependency/common_js_export_scanner.rs:188]: A little hack, webpack will skip rest visiting if found a `CommonjsExportRequireDependency`, but we always generate that dependency, need parser hook.  https://github.com/webpack/webpack/blob/1f99ad6367f2b8a6ef17cce0e058f7a67fb7db18/lib/dependencies/CommonJsExportsParserPlugin.js#L207-L219

**Review by ahabhgk — APPROVED** [review]

**Pull request merged**